### PR TITLE
Partially shared volumes hide data from databrary community - backend

### PR DESCRIPTION
**Pull request by kanishka-azimi**
DO NOT MERGE . (Ready for feedback; needs some manual testing and a frontend PR merged soon after).

These changes finish up an earlier sweeping change to volume sharing levels. The original change modified volume sharing levels from 
1. private
2. public 

to 

1. private
2. partially shared
3. public

When partially shared, all volume data besides overview information is hidden to public viewers as well as databrary community member viewers. The original implementation only hid data from public viewers. These changes implement hiding from databrary community member viewers.

The core of this change can be observed through the (now poorly named) volumeIsPublicRestricted function. It observes the computed permission for the current user "volumePermissionPolicy" and decides whether data should be hidden or not. The tuple observed consists of a role ("public", "shared", "readers", "editors", ...) and a refinement of the role with a policy. For most roles, there is only the default policy. For the "public" and "shared" roles, there is a default policy and a restricted policy.

The computed role/policy value for the current user on a given volume comes from multiple sources. The sources can be summarized as "direct access", "direct access via lab (authorized investigator of lab)", and "general group access". The role and policy for the current viewer is computed by taking the highest role/policy the user can achieve through the different sources. One is only allowed to annotate a role with a restricted policy when the role is being assigned through the "general group access" source.

A partially shared volume means the volume has been shared with the "public" group in the role "public" with policy "restricted", as well as being shared with the "databrary community" group in the role "shared" with policy "restricted". 

----

There is some noise from the refactor to use ```volumeRolePolicy``` . Its mostly syntactic, so it can be largely ignored.

### CODE REVIEW

File: schema/20180523-access-sharefull-contd.sql [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+        UNION ALL
+
+        SELECT mx.volume
+             , mx.child
+             , vap.result_perm
+             , vap.share_full
+        FROM vap_max AS mx
+            JOIN vap on (mx.volume, mx.mparent, mx.child) = (vap.volume, vap.parent, vap.child);
+
```
Comment (kanishka-azimi) [inline schema/20180523-access-sharefull-contd.sql]: The prior version of this query is in schema/20180523-access-sharefull.sql

File: src/Databrary/Controller/VolumeAccess.hs [hunk 2]
```diff
@@ -55,10 +55,13 @@ postVolumeAccess = action POST (pathAPI </> pathId </> pathVolumeAccessTarget) $
...
       then do
-        _ <- "share_full" .:> (deformCheck "Required" (not . (== FormDatumNone)) =<< deform) -- convulated way of requiring
-        Just <$> ("share_full" .:> deform)
+        if individual == PermissionSHARED -- TODO: remove me, temporary until front end updated
+        then pure (Just False)
+        else do
+          _ <- "share_full" .:> (deformCheck "Required" (not . (== FormDatumNone)) =<< deform) -- convulated way of requiring
+          Just <$> ("share_full" .:> deform)
       else pure Nothing 
     return a
```
Comment (kanishka-azimi) [inline src/Databrary/Controller/VolumeAccess.hs]: Deal with TODO as soon as front end is modified.

File: schema/20180523-access-sharefull-contd.sql [hunk 3]
```diff
@@ -0,0 +1,62 @@
+ALTER TABLE "volume_access" DROP CONSTRAINT "volume_access_check_full1";
+ALTER TABLE "volume_access" DROP CONSTRAINT "volume_access_check_full2";
+ALTER TABLE "volume_access" ADD CONSTRAINT "volume_access_check_full3"
+  CHECK ( party = -1 OR party = 0 OR share_full is null );
+-- expanding full2 for party 0 would be too complicated
```
Comment (kanishka-azimi) [inline schema/20180523-access-sharefull-contd.sql]: One can only use the policy "share_full" when the party is nobody/public (-1) or databrary site (0)

File: src/Databrary/Model/Permission/Types.hs [hunk 4]
```diff
@@ -111,7 +111,7 @@ instance Monoid Access where
 
 deriveLiftMany [''Permission, ''Access]
 
-data VolumeAccessPolicy = PublicRestricted | PermLevelDefault
+data VolumeAccessPolicy = PublicRestricted | SharedRestricted | PermLevelDefault
   deriving (Show, Eq)
```
Comment (kanishka-azimi) [inline src/Databrary/Model/Permission/Types.hs]: The type I am trying to reach is:
data VolumeAccessRolePolicy = None | Public PublicPolicy | Shared SharedPolicy | Read | Edit | Admin
For now, it mixes PublicPolicy and SharedPolicy together.

File: src/Databrary/Model/Volume.hs [hunk 5]
```diff
@@ -160,6 +161,7 @@ volumePublicAccessSummary vas =
              case accessPolicy of
                PublicRestricted -> PublicAccessRestricted
                PermLevelDefault -> PublicAccessFull
+               _ -> PublicAccessFull -- should never happen
```
Comment (kanishka-azimi) [inline src/Databrary/Model/Volume.hs]: Once I reach the ideal type above, this case will no longer be needed

File: schema/20180523-access-sharefull-contd.sql [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        UNION ALL
+
+        SELECT mx.volume
+             , mx.child
+             , vap.result_perm
+             , vap.share_full
+        FROM vap_max AS mx
+            JOIN vap on (mx.volume, mx.mparent, mx.child) = (vap.volume, vap.parent, vap.child);
+
```
Comment (kanishka-azimi) [inline schema/20180523-access-sharefull-contd.sql]: Computing the maximum in two steps instead of a single group by allows me to avoid the BOOL_AND hack and also makes sure the "share_full" value selected matches the maximum permission entry properly.

File: src/Databrary/View/Transcode.hs [hunk 7]
```diff
@@ -48,7 +48,7 @@ htmlTranscodes tl req = htmlTemplate req (Just "transcodes") $ \js -> do
           H.string $ show $ assetId $ assetRow $ transcodeAsset t
         H.td $ foldMap (H.string . show) transcodeStart
         H.td $ do
-          let p = view transcodeOwner
```
Comment (kanishka-azimi) [inline src/Databrary/View/Transcode.hs]: ignore

File: src/Databrary/Model/VolumeAccess/Types.hs [hunk 8]
```diff
@@ -26,8 +26,8 @@ getShareFullDefault targetParty individualAccessLevel =
     then nobodyPublicLegacyDefault
     else generalDefault
   where
-    getPartyId :: Party -> Id Party
-    getPartyId = partyId . partyRow
```
Comment (kanishka-azimi) [inline src/Databrary/Model/VolumeAccess/Types.hs]: ignore

File: src/Databrary/Model/Authorize.hs [hunk 9]
```diff
@@ -95,7 +95,7 @@ lookupAuthorization child parent
   | partyId (partyRow child) == partyId (partyRow parent) = return $ authorization $ selfAuthorize child
   | otherwise = do
     auth <- peek
-    if partyId (view auth) == partyId (partyRow child) && partyId (partyRow parent) == partyId (partyRow rootParty)
```
Comment (kanishka-azimi) [inline src/Databrary/Model/Authorize.hs]: ignore

File: schema/20180523-access-sharefull-contd.sql [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        SELECT volume, party, individual, share_full FROM volume_access
+
+        UNION ALL
+
+        SELECT mx.volume
+             , mx.child
+             , vap.result_perm
+             , vap.share_full
+        FROM vap_max AS mx
+            JOIN vap on (mx.volume, mx.mparent, mx.child) = (vap.volume, vap.parent, vap.child);
```
Comment (kanishka-azimi) [inline schema/20180523-access-sharefull-contd.sql]: If this query and authorize_inherit/authorize_view moved to logic in haskell later on, that would help with testability and being able to annotate each transformation with rich types and expressive names.

File: test/Databrary/Model/PermissionTest.hs [hunk 11]
```diff
@@ -21,3 +22,55 @@ unit_accessMember = do
     -- accessMember (TestContext { ctxIdentity = mkSession ... }) @?= PermissionNONE
     -- edge case
     accessMember (TestContext { ctxIdentity = IdentityNotNeeded }) @?= PermissionNONE
+
+runDataPermission4 :: (Release, Release) -> VolumeRolePolicy -> Permission
+runDataPermission4 (relPub, relPriv) rolePolicy =
+  dataPermission4 (const (EffectiveRelease {effRelPublic = relPub, effRelPrivate = relPriv})) (const rolePolicy) ()
```
Comment (kanishka-azimi) [inline test/Databrary/Model/PermissionTest.hs]: Maybe implement a helper function and test that instead; or use the Has pattern and change signature for dataPermission4 to (HasEffectiveRelease a) -> (HasRolePolicy b) -> ... , to cut down on boilerplate run function here.
